NitroxClient/MonoBehaviours/DiscordRP/DiscordRPController.cs

```diff
@@ -0,0 +1,153 @@
+
+        public void SpectateCallback(string secret) => Log.Info($"[Discord] Spectate: {secret}");
+
+        public void JoinCallback(string secret)
+        {
+            Log.Info($"[Discord] Join: {secret}");
+            if (SceneManager.GetActiveScene().name == "StartScreen")
+            {
+                MainMenuMultiplayerPanel.Main.OpenJoinServerMenu(secret);
+            }
+            else
+            {
+                Log.InGame("Please enter the main-menu if you want to join a session.");
+                Log.Warn("[Discord] Warn: Can't join a server outside of the main-menu.");
+            }
+        }
+
+        public void RequestCallback(ref DiscordRpc.DiscordUser request)
+        {
+            if (!ShowingWindow)
+            {
+                Log.Info($"[Discord] JoinRequest: Name:{request.username}#{request.discriminator} UserID:{request.userId}");
+                DiscordJoinRequestGui acceptRequest = gameObject.AddComponent<DiscordJoinRequestGui>();
+                acceptRequest.Request = request;
+                ShowingWindow = true;
+            }
+            else
+            {
+                Log.Debug("[Discord] Request window is already active.");
+            }
+        }
+
+        private void Awake()
+        {
+            DontDestroyOnLoad(gameObject);
+        }
+
+        private void Update()
+        {
+            DiscordRpc.RunCallbacks();
+        }
+
+        private void OnEnable()
+        {
+            Log.Info("[Discord] Init");
+            handlers = new DiscordRpc.EventHandlers
+            {
+                readyCallback = ReadyCallback
+            };
+            handlers.disconnectedCallback += DisconnectedCallback;
+            handlers.errorCallback += ErrorCallback;
+            handlers.joinCallback += JoinCallback;
+            handlers.spectateCallback += SpectateCallback;
+            handlers.requestCallback += RequestCallback;
+            DiscordRpc.Initialize(APPLICATION_ID, ref handlers, true, "");
+        }
+
+        private void OnDisable()
+        {
+            Log.Info("[Discord] Shutdown");
+            DiscordRpc.Shutdown();
+        }
+
+        public static void InitDiscordRPInGame(string username, int playerCount, string ipAddressPort)
+        {
+            Main.Presence.state = "In game";
+            Main.Presence.details = "Playing as " + username;
+            Main.Presence.startTimestamp = 0;
+            Main.Presence.partyId = "PartyID:"+CheckIP(ipAddressPort);
+            Main.Presence.partySize = playerCount;
+            Main.Presence.partyMax = 100;
+            Main.Presence.joinSecret = CheckIP(ipAddressPort);
+            Main.SendDiscordRP();
+        }
+
+        public static void InitDiscordRPMenu()
+        {
+            Main.Presence.state = "In menu";
+            Main.SendDiscordRP();
+        }
+
+        public static void UpdateDiscordRPPlayerCount(int playerCount)
```

Comment:
Instead of making these all static, should the caller reference DiscordRP.Main?  I believe this is the existing pattern that subnautica uses.